{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5a0be95c-8962-4e31-8d95-3a1c164ae612",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np \n",
    "import pandas as pd "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e289f738-9706-44b6-85fd-821aa80159dc",
   "metadata": {},
   "source": [
    "## Creating Map File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "60b6e011-c827-4b0b-8e17-be68a748e4a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chm</th>\n",
       "      <th>spos</th>\n",
       "      <th>epos</th>\n",
       "      <th>sgpos</th>\n",
       "      <th>egpos</th>\n",
       "      <th>n snps</th>\n",
       "      <th>S100.0</th>\n",
       "      <th>S100.1</th>\n",
       "      <th>S101.0</th>\n",
       "      <th>S101.1</th>\n",
       "      <th>...</th>\n",
       "      <th>LP6005592-DNA_D04.0</th>\n",
       "      <th>LP6005592-DNA_D04.1</th>\n",
       "      <th>LP6005519-DNA_C03.0</th>\n",
       "      <th>LP6005519-DNA_C03.1</th>\n",
       "      <th>LP6005519-DNA_D03.0</th>\n",
       "      <th>LP6005519-DNA_D03.1</th>\n",
       "      <th>LP6005677-DNA_C03.0</th>\n",
       "      <th>LP6005677-DNA_C03.1</th>\n",
       "      <th>LP6005677-DNA_A03.0</th>\n",
       "      <th>LP6005677-DNA_A03.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr16</td>\n",
       "      <td>33626</td>\n",
       "      <td>384376</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.38</td>\n",
       "      <td>967</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr16</td>\n",
       "      <td>384376</td>\n",
       "      <td>542174</td>\n",
       "      <td>0.38</td>\n",
       "      <td>0.59</td>\n",
       "      <td>330</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr16</td>\n",
       "      <td>542174</td>\n",
       "      <td>742038</td>\n",
       "      <td>0.59</td>\n",
       "      <td>0.86</td>\n",
       "      <td>405</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16</td>\n",
       "      <td>742038</td>\n",
       "      <td>846516</td>\n",
       "      <td>0.86</td>\n",
       "      <td>1.41</td>\n",
       "      <td>480</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr16</td>\n",
       "      <td>846516</td>\n",
       "      <td>897962</td>\n",
       "      <td>1.41</td>\n",
       "      <td>1.62</td>\n",
       "      <td>145</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715</th>\n",
       "      <td>chr16</td>\n",
       "      <td>89572818</td>\n",
       "      <td>89586409</td>\n",
       "      <td>133.63</td>\n",
       "      <td>133.67</td>\n",
       "      <td>40</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>716</th>\n",
       "      <td>chr16</td>\n",
       "      <td>89586409</td>\n",
       "      <td>89628638</td>\n",
       "      <td>133.67</td>\n",
       "      <td>133.74</td>\n",
       "      <td>185</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>717</th>\n",
       "      <td>chr16</td>\n",
       "      <td>89628638</td>\n",
       "      <td>89683139</td>\n",
       "      <td>133.74</td>\n",
       "      <td>133.78</td>\n",
       "      <td>235</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>718</th>\n",
       "      <td>chr16</td>\n",
       "      <td>89683139</td>\n",
       "      <td>89717130</td>\n",
       "      <td>133.78</td>\n",
       "      <td>133.80</td>\n",
       "      <td>110</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>719</th>\n",
       "      <td>chr16</td>\n",
       "      <td>89717130</td>\n",
       "      <td>90106346</td>\n",
       "      <td>133.80</td>\n",
       "      <td>134.05</td>\n",
       "      <td>1583</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>720 rows Ã— 8396 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      #chm      spos      epos   sgpos   egpos  n snps  S100.0  S100.1  \\\n",
       "0    chr16     33626    384376    0.00    0.38     967       2       2   \n",
       "1    chr16    384376    542174    0.38    0.59     330       2       2   \n",
       "2    chr16    542174    742038    0.59    0.86     405       2       2   \n",
       "3    chr16    742038    846516    0.86    1.41     480       2       2   \n",
       "4    chr16    846516    897962    1.41    1.62     145       2       2   \n",
       "..     ...       ...       ...     ...     ...     ...     ...     ...   \n",
       "715  chr16  89572818  89586409  133.63  133.67      40       2       6   \n",
       "716  chr16  89586409  89628638  133.67  133.74     185       2       6   \n",
       "717  chr16  89628638  89683139  133.74  133.78     235       2       6   \n",
       "718  chr16  89683139  89717130  133.78  133.80     110       2       6   \n",
       "719  chr16  89717130  90106346  133.80  134.05    1583       2       6   \n",
       "\n",
       "     S101.0  S101.1  ...  LP6005592-DNA_D04.0  LP6005592-DNA_D04.1  \\\n",
       "0         2       2  ...                    6                    6   \n",
       "1         2       2  ...                    6                    6   \n",
       "2         2       2  ...                    6                    6   \n",
       "3         2       2  ...                    6                    6   \n",
       "4         2       2  ...                    6                    6   \n",
       "..      ...     ...  ...                  ...                  ...   \n",
       "715       3       3  ...                    6                    6   \n",
       "716       3       3  ...                    6                    6   \n",
       "717       3       3  ...                    6                    6   \n",
       "718       3       3  ...                    6                    6   \n",
       "719       3       3  ...                    6                    6   \n",
       "\n",
       "     LP6005519-DNA_C03.0  LP6005519-DNA_C03.1  LP6005519-DNA_D03.0  \\\n",
       "0                      5                    2                    2   \n",
       "1                      5                    2                    2   \n",
       "2                      5                    2                    2   \n",
       "3                      5                    2                    2   \n",
       "4                      5                    2                    2   \n",
       "..                   ...                  ...                  ...   \n",
       "715                    6                    5                    2   \n",
       "716                    6                    5                    2   \n",
       "717                    6                    5                    2   \n",
       "718                    6                    5                    2   \n",
       "719                    6                    5                    2   \n",
       "\n",
       "     LP6005519-DNA_D03.1  LP6005677-DNA_C03.0  LP6005677-DNA_C03.1  \\\n",
       "0                      2                    6                    2   \n",
       "1                      2                    6                    2   \n",
       "2                      2                    6                    2   \n",
       "3                      2                    6                    2   \n",
       "4                      2                    6                    2   \n",
       "..                   ...                  ...                  ...   \n",
       "715                    2                    2                    2   \n",
       "716                    2                    2                    2   \n",
       "717                    2                    2                    2   \n",
       "718                    2                    2                    2   \n",
       "719                    2                    2                    2   \n",
       "\n",
       "     LP6005677-DNA_A03.0  LP6005677-DNA_A03.1  \n",
       "0                      2                    2  \n",
       "1                      2                    2  \n",
       "2                      2                    2  \n",
       "3                      2                    2  \n",
       "4                      2                    2  \n",
       "..                   ...                  ...  \n",
       "715                    2                    2  \n",
       "716                    2                    2  \n",
       "717                    2                    2  \n",
       "718                    2                    2  \n",
       "719                    2                    2  \n",
       "\n",
       "[720 rows x 8396 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "example_msp_file = pd.read_csv(\"ancestry_results/covid_ref_chr16_rfmix.msp.tsv\", sep=\"\\t\", skiprows=1)\n",
    "example_msp_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3b151076-0bb3-4dac-a4d4-2ba6858493b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_oceanic():\n",
    "    for chr in np.arange(1, 23):\n",
    "        file = pd.read_csv(\"ancestry_results/covid_ref_chr{}_rfmix.msp.tsv\".format(chr), sep=\"\\t\", skiprows=1)\n",
    "        new_file = file.copy()\n",
    "        new_file['N319.0'] = (np.ones(len(file.index))*4).astype(int)\n",
    "        new_file['N319.1'] = (np.ones(len(file.index))*4).astype(int)\n",
    "        new_file.to_csv(\"ancestry_results_with_oceanic/covid_ref_chr{}_rfmix.msp.tsv\".format(chr), index=False, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "a23b9375-9b44-4e7a-8663-f3ce64858db1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_map_file(path1, path2):\n",
    "    add_oceanic()\n",
    "    map_file = pd.DataFrame(columns = [\"cM\", \"chr\"])\n",
    "    for chr in np.arange(1, 23):\n",
    "        file_path = path1+str(chr)+path2\n",
    "        chr_file = pd.read_csv(file_path, sep=\"\\t\")\n",
    "        sixes2twos_msp(chr, chr_file)\n",
    "        chr_map = chr_file[[\"sgpos\", \"#chm\"]].rename(columns={\"sgpos\":\"cM\", \"#chm\":\"chr\"})\n",
    "        chr_map['chr'] = chr_map.apply(lambda x: x['chr'][3:], 1) \n",
    "        map_file = pd.concat([map_file, chr_map])\n",
    "        print(\"chr {} map completed\".format(chr))\n",
    "    map_file.to_csv(\"map_file.tsv\", index=False, sep=\"\\t\")\n",
    "\n",
    "def sixes2twos_msp(chr, chr_file):\n",
    "    cols = [col for col in chr_file.columns if col == \"sgpos\" or col.startswith(\"S\") and not col.startswith(\"SS\")\n",
    "            or col.startswith(\"N\") and not col.startswith(\"NA\")]\n",
    "    new_df = chr_file[cols]\n",
    "    participant_cols = [col for col in cols if col != \"sgpos\"]\n",
    "    new_cleaned_df = new_df.copy()\n",
    "    sixes2twos = new_df.loc[:, participant_cols].mask(new_df[participant_cols] == 6, 2)\n",
    "    new_cleaned_df.loc[:, participant_cols] = sixes2twos\n",
    "    new_cleaned_df.to_csv(\"cleaned_ancestry_results/chr{}.msp.tsv\".format(chr), index=False, sep=\"\\t\")\n",
    "    print(\"chr {} cleaning completed\".format(chr))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "8148be70-2e7a-4a0f-9476-63f2119acda5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr 1 cleaning completed\n",
      "chr 1 map completed\n",
      "chr 2 cleaning completed\n",
      "chr 2 map completed\n",
      "chr 3 cleaning completed\n",
      "chr 3 map completed\n",
      "chr 4 cleaning completed\n",
      "chr 4 map completed\n",
      "chr 5 cleaning completed\n",
      "chr 5 map completed\n",
      "chr 6 cleaning completed\n",
      "chr 6 map completed\n",
      "chr 7 cleaning completed\n",
      "chr 7 map completed\n",
      "chr 8 cleaning completed\n",
      "chr 8 map completed\n",
      "chr 9 cleaning completed\n",
      "chr 9 map completed\n",
      "chr 10 cleaning completed\n",
      "chr 10 map completed\n",
      "chr 11 cleaning completed\n",
      "chr 11 map completed\n",
      "chr 12 cleaning completed\n",
      "chr 12 map completed\n",
      "chr 13 cleaning completed\n",
      "chr 13 map completed\n",
      "chr 14 cleaning completed\n",
      "chr 14 map completed\n",
      "chr 15 cleaning completed\n",
      "chr 15 map completed\n",
      "chr 16 cleaning completed\n",
      "chr 16 map completed\n",
      "chr 17 cleaning completed\n",
      "chr 17 map completed\n",
      "chr 18 cleaning completed\n",
      "chr 18 map completed\n",
      "chr 19 cleaning completed\n",
      "chr 19 map completed\n",
      "chr 20 cleaning completed\n",
      "chr 20 map completed\n",
      "chr 21 cleaning completed\n",
      "chr 21 map completed\n",
      "chr 22 cleaning completed\n",
      "chr 22 map completed\n"
     ]
    }
   ],
   "source": [
    "map_file = create_map_file(\"ancestry_results_with_oceanic/covid_ref_chr\", \"_rfmix.msp.tsv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b87dde07-4753-49c0-bc4d-55817d56864b",
   "metadata": {},
   "source": [
    "## Converting to VCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6a1edc7e-04df-4be0-96bb-cbedae45a85a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_map_file(path1, path2):\n",
    "    #add_oceanic()\n",
    "    map_file = pd.DataFrame(columns = [\"position\", \"chr\"])\n",
    "    for chr in np.arange(1, 23):\n",
    "        file_path = path1+str(chr)+path2\n",
    "        chr_file = pd.read_csv(file_path, sep=\"\\t\")\n",
    "        sixes2twos_msp(chr, chr_file)\n",
    "        chr_map = chr_file[[\"spos\", \"#chm\"]].rename(columns={\"spos\":\"position\", \"#chm\":\"chr\"})\n",
    "        \n",
    "def sixes2twos_msp(chr, chr_file):\n",
    "    cols = [col for col in chr_file.columns if col == \"spos\" or col.startswith(\"S\") and not col.startswith(\"SS\")\n",
    "            or col.startswith(\"N\") and not col.startswith(\"NA\")]\n",
    "    new_df = chr_file[cols]\n",
    "    participant_cols = [col for col in cols if col != \"spos\"]\n",
    "    cleaned_pre_vcf = new_df.copy()\n",
    "    sixes2twos = new_df.loc[:, participant_cols].mask(new_df[participant_cols] == 6, 2)\n",
    "    cleaned_pre_vcf.loc[:, participant_cols] = sixes2twos\n",
    "    cleaned_pre_vcf.to_csv(\"cleaned_pre_vcf/chr{}.msp.tsv\".format(chr), index=False, sep=\"\\t\")\n",
    "    print(\"chr {} cleaning completed\".format(chr))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "305b5675-3a36-446a-8754-27dee390e46c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr 1 cleaning completed\n",
      "chr 2 cleaning completed\n",
      "chr 3 cleaning completed\n",
      "chr 4 cleaning completed\n",
      "chr 5 cleaning completed\n",
      "chr 6 cleaning completed\n",
      "chr 7 cleaning completed\n",
      "chr 8 cleaning completed\n",
      "chr 9 cleaning completed\n",
      "chr 10 cleaning completed\n",
      "chr 11 cleaning completed\n",
      "chr 12 cleaning completed\n",
      "chr 13 cleaning completed\n",
      "chr 14 cleaning completed\n",
      "chr 15 cleaning completed\n",
      "chr 16 cleaning completed\n",
      "chr 17 cleaning completed\n",
      "chr 18 cleaning completed\n",
      "chr 19 cleaning completed\n",
      "chr 20 cleaning completed\n",
      "chr 21 cleaning completed\n",
      "chr 22 cleaning completed\n"
     ]
    }
   ],
   "source": [
    "create_map_file(\"ancestry_results_with_oceanic/covid_ref_chr\", \"_rfmix.msp.tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "86df02d9-0ca4-439f-9b21-f41cb865897c",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Converts .msp file to .vcf format\n",
    "###\n",
    "### msp_file: the original file ending in .msp\n",
    "### output_file: file path and name for the resulting .vcf file\n",
    "### ancestry: integer representing the ancestry of interest to binary encode the data\n",
    "### EAS=1, EUR/WAS=2, NAT=3, OCE=4, SAS=5\n",
    "\n",
    "def msp2vcf(msp_file, output_file, ancestry):\n",
    "    msp = open(msp_file)\n",
    "    chromosome = msp_file[-9]\n",
    "    output = open(output_file, 'w')\n",
    "    output.write(\"##fileformat=VCFv4.2 \\n##filedate=20220513 \\n##source={}\\n\".format(msp_file))\n",
    "    output.write('##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\\n')\n",
    "    for line in msp:\n",
    "        bits = line.split()\n",
    "        # write the participant ID's to the file, joining their entries\n",
    "        # (eg. S100.0, S100.1 -> S100)\n",
    "        if bits[0] == \"spos\":\n",
    "            line_str = \"#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\"\n",
    "            line_str = \"\\t\".join([line_str, bits[1][:-2]])\n",
    "            for i in np.arange(3, len(bits)):\n",
    "                if (i%2 == 1):\n",
    "                    line_str = \"\\t\".join([line_str, bits[i][:-2]])\n",
    "        # write the binary coded ancestries for each participant\n",
    "        # for example, if ancestry=2, then 3 2 would be coded as 0|1 in the output file\n",
    "        else:\n",
    "            line_str = \"\\t\".join([chromosome, bits[0], '.', '.', '.', '.', 'PASS', '.', 'GT'])\n",
    "            for i in np.arange(1, len(bits)):\n",
    "                binary = int(int(bits[i]) == ancestry)\n",
    "                if (i%2 == 1):\n",
    "                    line_str += \"\\t\" + str(binary) + \"|\"\n",
    "                else:\n",
    "                    line_str += str(binary)\n",
    "        output.write(line_str + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "2247a7ff-cdf0-4e3d-a2d0-96166ec77972",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Converts all msp files to vcf format by calling msp2vcf()\n",
    "###\n",
    "### input_dir: string path to the directory with the .msp files\n",
    "### output_dir: string path to the directory in which to store .vcp files\n",
    "### ancestries: list of ancestry integers of interest (in our case, [1,2,3,4,5])\n",
    "\n",
    "def convert_all2vcf(input_dir, output_dir, ancestries):\n",
    "    for chrom in np.arange(1, 23):\n",
    "        for ancestry in ancestries:\n",
    "            msp_file = input_dir + \"chr{}.msp.tsv\".format(chrom)\n",
    "            output_file = output_dir + \"covid_ref_chr{}_ancestry{}_rfmix.vcf\".format(chrom, ancestry)\n",
    "            msp2vcf(msp_file, output_file, ancestry)\n",
    "            print(\"Converted msp->vcf for chromosome #{}, ancestry #{}\".format(chrom, ancestry))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "0f441282-fbd3-46b4-bdd1-d7aef90fc192",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted msp->vcf for chromosome #1, ancestry #1\n",
      "Converted msp->vcf for chromosome #1, ancestry #2\n",
      "Converted msp->vcf for chromosome #1, ancestry #3\n",
      "Converted msp->vcf for chromosome #1, ancestry #4\n",
      "Converted msp->vcf for chromosome #1, ancestry #5\n",
      "Converted msp->vcf for chromosome #2, ancestry #1\n",
      "Converted msp->vcf for chromosome #2, ancestry #2\n",
      "Converted msp->vcf for chromosome #2, ancestry #3\n",
      "Converted msp->vcf for chromosome #2, ancestry #4\n",
      "Converted msp->vcf for chromosome #2, ancestry #5\n",
      "Converted msp->vcf for chromosome #3, ancestry #1\n",
      "Converted msp->vcf for chromosome #3, ancestry #2\n",
      "Converted msp->vcf for chromosome #3, ancestry #3\n",
      "Converted msp->vcf for chromosome #3, ancestry #4\n",
      "Converted msp->vcf for chromosome #3, ancestry #5\n",
      "Converted msp->vcf for chromosome #4, ancestry #1\n",
      "Converted msp->vcf for chromosome #4, ancestry #2\n",
      "Converted msp->vcf for chromosome #4, ancestry #3\n",
      "Converted msp->vcf for chromosome #4, ancestry #4\n",
      "Converted msp->vcf for chromosome #4, ancestry #5\n",
      "Converted msp->vcf for chromosome #5, ancestry #1\n",
      "Converted msp->vcf for chromosome #5, ancestry #2\n",
      "Converted msp->vcf for chromosome #5, ancestry #3\n",
      "Converted msp->vcf for chromosome #5, ancestry #4\n",
      "Converted msp->vcf for chromosome #5, ancestry #5\n",
      "Converted msp->vcf for chromosome #6, ancestry #1\n",
      "Converted msp->vcf for chromosome #6, ancestry #2\n",
      "Converted msp->vcf for chromosome #6, ancestry #3\n",
      "Converted msp->vcf for chromosome #6, ancestry #4\n",
      "Converted msp->vcf for chromosome #6, ancestry #5\n",
      "Converted msp->vcf for chromosome #7, ancestry #1\n",
      "Converted msp->vcf for chromosome #7, ancestry #2\n",
      "Converted msp->vcf for chromosome #7, ancestry #3\n",
      "Converted msp->vcf for chromosome #7, ancestry #4\n",
      "Converted msp->vcf for chromosome #7, ancestry #5\n",
      "Converted msp->vcf for chromosome #8, ancestry #1\n",
      "Converted msp->vcf for chromosome #8, ancestry #2\n",
      "Converted msp->vcf for chromosome #8, ancestry #3\n",
      "Converted msp->vcf for chromosome #8, ancestry #4\n",
      "Converted msp->vcf for chromosome #8, ancestry #5\n",
      "Converted msp->vcf for chromosome #9, ancestry #1\n",
      "Converted msp->vcf for chromosome #9, ancestry #2\n",
      "Converted msp->vcf for chromosome #9, ancestry #3\n",
      "Converted msp->vcf for chromosome #9, ancestry #4\n",
      "Converted msp->vcf for chromosome #9, ancestry #5\n",
      "Converted msp->vcf for chromosome #10, ancestry #1\n",
      "Converted msp->vcf for chromosome #10, ancestry #2\n",
      "Converted msp->vcf for chromosome #10, ancestry #3\n",
      "Converted msp->vcf for chromosome #10, ancestry #4\n",
      "Converted msp->vcf for chromosome #10, ancestry #5\n",
      "Converted msp->vcf for chromosome #11, ancestry #1\n",
      "Converted msp->vcf for chromosome #11, ancestry #2\n",
      "Converted msp->vcf for chromosome #11, ancestry #3\n",
      "Converted msp->vcf for chromosome #11, ancestry #4\n",
      "Converted msp->vcf for chromosome #11, ancestry #5\n",
      "Converted msp->vcf for chromosome #12, ancestry #1\n",
      "Converted msp->vcf for chromosome #12, ancestry #2\n",
      "Converted msp->vcf for chromosome #12, ancestry #3\n",
      "Converted msp->vcf for chromosome #12, ancestry #4\n",
      "Converted msp->vcf for chromosome #12, ancestry #5\n",
      "Converted msp->vcf for chromosome #13, ancestry #1\n",
      "Converted msp->vcf for chromosome #13, ancestry #2\n",
      "Converted msp->vcf for chromosome #13, ancestry #3\n",
      "Converted msp->vcf for chromosome #13, ancestry #4\n",
      "Converted msp->vcf for chromosome #13, ancestry #5\n",
      "Converted msp->vcf for chromosome #14, ancestry #1\n",
      "Converted msp->vcf for chromosome #14, ancestry #2\n",
      "Converted msp->vcf for chromosome #14, ancestry #3\n",
      "Converted msp->vcf for chromosome #14, ancestry #4\n",
      "Converted msp->vcf for chromosome #14, ancestry #5\n",
      "Converted msp->vcf for chromosome #15, ancestry #1\n",
      "Converted msp->vcf for chromosome #15, ancestry #2\n",
      "Converted msp->vcf for chromosome #15, ancestry #3\n",
      "Converted msp->vcf for chromosome #15, ancestry #4\n",
      "Converted msp->vcf for chromosome #15, ancestry #5\n",
      "Converted msp->vcf for chromosome #16, ancestry #1\n",
      "Converted msp->vcf for chromosome #16, ancestry #2\n",
      "Converted msp->vcf for chromosome #16, ancestry #3\n",
      "Converted msp->vcf for chromosome #16, ancestry #4\n",
      "Converted msp->vcf for chromosome #16, ancestry #5\n",
      "Converted msp->vcf for chromosome #17, ancestry #1\n",
      "Converted msp->vcf for chromosome #17, ancestry #2\n",
      "Converted msp->vcf for chromosome #17, ancestry #3\n",
      "Converted msp->vcf for chromosome #17, ancestry #4\n",
      "Converted msp->vcf for chromosome #17, ancestry #5\n",
      "Converted msp->vcf for chromosome #18, ancestry #1\n",
      "Converted msp->vcf for chromosome #18, ancestry #2\n",
      "Converted msp->vcf for chromosome #18, ancestry #3\n",
      "Converted msp->vcf for chromosome #18, ancestry #4\n",
      "Converted msp->vcf for chromosome #18, ancestry #5\n",
      "Converted msp->vcf for chromosome #19, ancestry #1\n",
      "Converted msp->vcf for chromosome #19, ancestry #2\n",
      "Converted msp->vcf for chromosome #19, ancestry #3\n",
      "Converted msp->vcf for chromosome #19, ancestry #4\n",
      "Converted msp->vcf for chromosome #19, ancestry #5\n",
      "Converted msp->vcf for chromosome #20, ancestry #1\n",
      "Converted msp->vcf for chromosome #20, ancestry #2\n",
      "Converted msp->vcf for chromosome #20, ancestry #3\n",
      "Converted msp->vcf for chromosome #20, ancestry #4\n",
      "Converted msp->vcf for chromosome #20, ancestry #5\n",
      "Converted msp->vcf for chromosome #21, ancestry #1\n",
      "Converted msp->vcf for chromosome #21, ancestry #2\n",
      "Converted msp->vcf for chromosome #21, ancestry #3\n",
      "Converted msp->vcf for chromosome #21, ancestry #4\n",
      "Converted msp->vcf for chromosome #21, ancestry #5\n",
      "Converted msp->vcf for chromosome #22, ancestry #1\n",
      "Converted msp->vcf for chromosome #22, ancestry #2\n",
      "Converted msp->vcf for chromosome #22, ancestry #3\n",
      "Converted msp->vcf for chromosome #22, ancestry #4\n",
      "Converted msp->vcf for chromosome #22, ancestry #5\n"
     ]
    }
   ],
   "source": [
    "convert_all2vcf(\"cleaned_pre_vcf/\", \"vcf_files/\", [1,2,3,4,5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6700c0a-b852-4fbc-87c3-efef313c148c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!python analysis_pipeline/vcf2gds.py tester.config --cluster_type Slurm_Cluster --cluster_file analysis_pipeline/sherlock_cluster_sy.json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cab10a3-cbea-4c11-8d68-c98f4e249860",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ca8d16b-2211-4d4c-aa1e-7008118228fb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
